pull: special case all-tags semantics

vrothberg opened the pull request:
Supporting the all-tags semantics added some non-trivial code to the
pull command which does not make use of `registries.conf` and introduced
some regressions such as not adhering to the configured search registries.

Speacial case the all-tags flags to let existing users of all-tags
continue working while others can work again.  This implies that the
all-tags pull does not adhere to configured search registries while the
default (non-all-tags) pull does.

Note that this is a purely symptomaic fix.  A final solution should
include Buildah and the c/image library to avoid redundant and
error-prone code across the projects.

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1701922
Signed-off-by: Valentin Rothberg <rothberg@redhat.com>

vrothberg commented:
Fix for https://bugzilla.redhat.com/show_bug.cgi?id=1701922

@rhatdan @mheon @baude @TomSweeneyRedHat @mtrmac PTAL

My initial plan was to wire the all-tags support into the libpod backend but that seemed to be non-trivial. Moreover, we should aim for consolidating the code with buildah (and c/image?). As buildah cannot vendor in libpod, working on the libpod backend did not seem reasonable to me.

The idea of this PR is to special-case the all-tags flag until we have plan.

mtrmac commented on cmd/podman/pull.go:
```diff
@@ -122,68 +123,81 @@ func pullCmd(c *cliconfig.PullValues) (retError error) {
-		if len(names) > 1 {
-			fmt.Println("Pulled Images:")
+		imgID := newImage.ID()
+		fmt.Println(imgID)
```
(Absolutely non-blocking: I wouldn’t bother with the `imgID` variable.)

vrothberg replied:
Good catch, fixed.

vrothberg commented:
@mtrmac, thanks for the review. I fixed the nit and updated the commit message to a) be more explicit about the semantic consequences of the change and b) to also link to the BZ.

TomSweeneyRedHat commented:
I don't see a man page change here for --all-tags and think we should at least note there the difference in the registries search.

vrothberg commented:
> I don't see a man page change here for --all-tags and think we should at least note there the difference in the registries search.

Good catch, thanks! I added a sentence.

> Changes look reasonable to me, but the tests aren't hip. Having the all-tags live somewhere in c/image makes sense to me if it's doable, that way both Buildah and Libpod could make use of it.

The failures are caused by flakes and the runc issue (needs to be updated). Not sure when it's getting fixed.

TomSweeneyRedHat commented on docs/podman-pull.1.md:
```diff
@@ -49,6 +49,8 @@ Image stored in local container/storage
 All tagged images in the repository will be pulled.
 
+Note: When using the all-tags flag, Podman will not iterate over the search registries in the containers-registries.conf(5) but will always use docker.io for unqualified image names.
+
```
:+1:

TomSweeneyRedHat commented:
LGTM and tests well.  Now if we can get the CI to behave.  Thanks @vrothberg !

mheon commented:
LGTM once we get the new containers-selinux into the CI images and things go green

rhatdan commented:
Needs a rebase to get the SELinux fixes.

vrothberg commented:
> Needs a rebase to get the SELinux fixes.

Done :heavy_check_mark:

baude commented:
/lgtm
